### PR TITLE
plan: add push topn optimizing.

### PR DESCRIPTION
**Pull request by hanfei1991**
we push topn firstly to remove the `limit` property from `required property`.
@shenli @coocood @zimulala PTAL

### CODE REVIEW

File: plan/plan.go [hunk 1]
```diff
@@ -191,6 +191,8 @@ type LogicalPlan interface {
 
 	// buildKeyInfo will collect the information of unique keys into schema.
 	buildKeyInfo()
+
+	pushTopN(topn *Sort) LogicalPlan
```
Comment (ngaut) [inline plan/plan.go]: s/push/pushdown/

**Comment by shenli** [comment]
CI failed.

File: executor/explain_test.go [hunk 2]
```diff
@@ -350,10 +350,10 @@ func (s *testSuite) TestExplain(c *C) {
 		{
 			"select * from t2 order by t2.c2 limit 0, 1",
 			[]string{
-				"TableScan_5", "Sort_6",
```
Comment (shenli) [inline executor/explain_test.go]: DO NOT forget update mysql-test.

File: plan/push_topn.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+)
+
+type pushTopNSolver struct {
```
Comment (shenli) [inline plan/push_topn.go]: Please add a few comment for this file.

File: executor/explain_test.go (hunk 2 above)
Comment (hanfei1991) [inline executor/explain_test.go]: ok

File: plan/push_topn.go (hunk 3 above)
Comment (hanfei1991) [inline plan/push_topn.go]: added , PTAL

File: plan/push_topn.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+		p.SetParents(s)
+		s.SetSchema(p.Schema().Clone())
+		return s
+	}
+}
+
+func (s *Sort) pushDownTopN(topN *Sort) LogicalPlan {
+	if topN.isLimit() {
+		s.ExecLimit = topN.ExecLimit
+		// If a Limit is pushDowned, the Sort should be converted to topN and be pushed again.
```
Comment (coocood) [inline plan/push_topn.go]: is pushed down.

File: plan/plan.go [hunk 5]
```diff
@@ -128,6 +128,8 @@ type LogicalPlan interface {
 
 	// buildKeyInfo will collect the information of unique keys into schema.
 	buildKeyInfo()
+
+	pushDownTopN(topn *Sort) LogicalPlan
```
Comment (coocood) [inline plan/plan.go]: This interface contains a specific type in the argument, which is an anti-pattern.

File: executor/explain_test.go [hunk 6]
```diff
@@ -637,7 +637,12 @@ func (s *testSuite) TestExplain(c *C) {
 				`{
     "limit": 10,
     "offset": 0,
-    "child": "TableScan_7"
+    "child": "TableScan_13"
+}`,
+				`{
```
Comment (coocood) [inline executor/explain_test.go]: The limit is duplicated.

**Comment by coocood** [comment]
If the code of old limit push down is not removed, we can not verify this code is correct.

File: plan/push_topn.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+)
+
+// pushDownTopNSolver push down the topn or limit. In future we will remove the limit from `requiredProperty` in CBO phase.
```
Comment (zimulala) [inline plan/push_topn.go]: s/push/pushes

File: plan/push_topn.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+func (s *Sort) setChild(p LogicalPlan) LogicalPlan {
+	if s.isEmpty() {
+		return p
+	} else if s.isLimit() {
+		limit := Limit{Count: s.ExecLimit.Count, Offset: s.ExecLimit.Offset}.init(s.allocator, s.ctx)
+		limit.SetChildren(p)
+		p.SetParents(limit)
+		limit.SetSchema(p.Schema().Clone())
+		return limit
+	} else {
```
Comment (zimulala) [inline plan/push_topn.go]: This code doesn't need to add `else`.

File: plan/push_topn.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+func (s *Sort) pushDownTopN(topN *Sort) LogicalPlan {
+	if topN.isLimit() {
+		s.ExecLimit = topN.ExecLimit
+		// If a Limit is pushed down, the Sort should be converted to topN and be pushed again.
+		return s.children[0].(LogicalPlan).pushDownTopN(s)
+	} else if topN.isEmpty() {
+		// If nothing is pushDowned, just continue to push nothing to its child.
+		return s.baseLogicalPlan.pushDownTopN(topN)
+	} else {
```
Comment (zimulala) [inline plan/push_topn.go]: Ditto.

File: plan/logical_plan_test.go [hunk 10]
```diff
@@ -1632,3 +1632,106 @@ func checkVisitInfo(c *C, v1, v2 []visitInfo, comment CommentInterface) {
...
+		},
+		// Test Limit + Join + Proj.
+		{
+			sql:  "select * from t, t s limit 5",
+			best: "Join{DataScan(t)->DataScan(s)}->Limit->Projection",
+		},
+		// Test TopN + Left Join + Proj.
+		{
+			sql:  "select * from t left outer join t s on t.a = s.a order by t.a limit 5",
+			best: "Join{DataScan(t)->Sort + Limit(5) + Offset(0)->DataScan(s)}(test.t.a,s.a)->Sort + Limit(5) + Offset(0)->Projection",
```
Comment (coocood) [inline plan/logical_plan_test.go]: The later `Sort + Limit(5) + Offset(0)` can be changed to `Limit(5) + Offset(0)`?

Comment (hanfei1991) [inline plan/logical_plan_test.go]: You cannot assume join can keep order, it may execute parallelly.

File: plan/logical_plan_test.go [hunk 11]
```diff
@@ -1632,3 +1632,106 @@ func checkVisitInfo(c *C, v1, v2 []visitInfo, comment CommentInterface) {
...
+		{
+			sql:  "select * from t union all (select * from t s) order by a,b limit 5",
+			best: "UnionAll{DataScan(t)->Sort + Limit(5) + Offset(0)->Projection->DataScan(s)->Sort + Limit(5) + Offset(0)->Projection}->Sort + Limit(5) + Offset(0)",
+		},
+		// Test Limit + UA + Proj + Sort.
+		{
+			sql:  "select * from t union all (select * from t s order by a) limit 5",
+			best: "UnionAll{DataScan(t)->Limit->Projection->DataScan(s)->Sort + Limit(5) + Offset(0)->Projection->Projection}->Limit",
+		},
+	}
```
Comment (coocood) [inline plan/logical_plan_test.go]: Need more test cases for Selection.

**Comment by coocood** [comment]
Why is `Selection` not implemented?

**Comment by hanfei1991** [comment]
@coocood baseLogicalPlan is a topN blocker.

**Comment by coocood** [comment]
When predicate is pushed down, topN can be pushed too.

**Comment by hanfei1991** [comment]
@coocood  We call predicate push down first

**Comment by coocood** [comment]
LGTM

File: plan/topn_push_down.go [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+package plan
+
+import (
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+)
+
+// pushDownTopNSolver pushes down the topN or limit. In future we will remove the limit from `requiredProperty` in CBO phase.
+type pushDownTopNSolver struct {
+	ctx       context.Context
```
Comment (shenli) [inline plan/topn_push_down.go]: The two members are not used.

File: plan/decorrelate.go [hunk 13]
```diff
@@ -95,6 +95,7 @@ func (s *decorrelateSolver) optimize(p LogicalPlan, _ context.Context, _ *idAllo
 			join := &apply.LogicalJoin
 			innerPlan.SetParents(join)
 			outerPlan.SetParents(join)
+			join.self = join
```
Comment (shenli) [inline plan/decorrelate.go]: This is a bug fix?

File: plan/topn_push_down.go [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+)
+
+// pushDownTopNSolver pushes down the topN or limit. In future we will remove the limit from `requiredProperty` in CBO phase.
+type pushDownTopNSolver struct {
```
Comment (shenli) [inline plan/topn_push_down.go]: pushDownTopNOptimizer is a better name.

File: plan/topn_push_down.go [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+)
+
+// pushDownTopNSolver pushes down the topN or limit. In future we will remove the limit from `requiredProperty` in CBO phase.
```
Comment (shenli) [inline plan/topn_push_down.go]: In the future

File: plan/decorrelate.go (hunk 13 above)
Comment (hanfei1991) [inline plan/decorrelate.go]: yes, the test has cover this

**Pull request merged**